### PR TITLE
refactor: query factory

### PR DESCRIPTION
**Pull request by DanielSchiavini**
- create query factory to remove some boiler plate code that is shared between all queries

image

The PR is basically about this change. Instead of creating a `getXQueryOptions`, `enabledX`, `XValidationGroup`, `XValidationSuite` and `keys.X` functions, all you need to do is `queryFactory()` and that will generate:

```typescript
    assertValidity: () => void
    checkValidity: () => void
    isEnabled: boolean
    queryKey: () => TKey
    getQueryOptions: () => QueryOptions
    getQueryData: (params) => TData
    useQuery: UseQueryHook
    invalidate: (params) => void
```

### CODE REVIEW

File: packages/curve-lib/src/shared/model/root-keys.ts
Comment (ignromanov) [inline packages/curve-lib/src/shared/model/root-keys.ts:1]: I would put it to the `shared/model/query/`

File: packages/curve-lib/src/shared/model/factory.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  TGroup extends string = string,
+  TCallback extends CB = CB<TQuery, TField[]>
+>({
+    query, queryParams, queryKey, staleTime, validationSuite, dependencies
+  }: QueryFactoryInput<TQuery, TKey, TData, TParams, TField, TGroup, TCallback>): QueryFactoryOutput<TQuery, TKey, TData, TParams> {
+
+  // todo: get rid of ValidatedData<T> use NonValidatedFields<T> instead
+  const assertValidity = (data: TParams, fields?: TField[]) => (sharedAssertValidity(validationSuite, data, fields) as unknown as TQuery)
+
+  const isEnabled = (params: TParams) => checkValidity(validationSuite, params) && !dependencies?.(params).find(key => !queryClient.getQueryData(key))
```
Comment (ignromanov) [inline packages/curve-lib/src/shared/model/factory.ts:25]:
```suggestion
  const isEnabled = (params: TParams) => checkValidity(validationSuite, params) && !dependencies?.(params). some(key => !queryClient.getQueryData(key))
```
looks better

File: apps/main/src/entities/gauge/model/query-options.ts [hunk 2]
```diff
@@ -10,54 +10,58 @@
...
 
-export const getGaugeStatusQueryOptions = (params: GaugeQueryParams) =>
-  queryOptions({
-    queryKey: keys.status(params),
-    queryFn: api.queryGaugeStatus,
-    staleTime: REFRESH_INTERVAL['5m'],
-    enabled: conditions.enabledGaugeStatus(params),
-  })
+export const gaugeStatus = queryFactory({
+  queryKey: ({ chainId, poolId }: GaugeParams) => ['chain', chainId, 'pool', poolId, 'gauge', 'status'] as const,
```
Comment (ignromanov) [inline apps/main/src/entities/gauge/model/query-options.ts:25]: How can we use this key from outside?
To pass it to `dependencies`  for example

Comment (DanielSchiavini) [reply apps/main/src/entities/gauge/model/query-options.ts:25]: `gaugeStatus.queryKey({...})`

Comment (DanielSchiavini) [reply apps/main/src/entities/gauge/model/query-options.ts:25]: It's [re-exported](https://github.com/curvefi/curve-frontend/blob/108fc68d87ce09855c08a80ad1810c0477f4bbbb/packages/curve-lib/src/shared/model/factory.ts#L43) from the [factory output](https://github.com/curvefi/curve-frontend/blob/108fc68d87ce09855c08a80ad1810c0477f4bbbb/packages/curve-lib/src/shared/types/factory.ts#L34)

**Review by ignromanov — DISMISSED** [review]

**Review by amytsang — DISMISSED** [review]

**Review by amytsang — DISMISSED** [review]

File: apps/main/src/entities/gauge/api/rewards-mutate.ts [hunk 3]
```diff
@@ -9,46 +9,24 @@
...
-> = async (queryKey) => {
-  const [, chainId, , poolId, , , rewardTokenId, distributorId] = queryKey
-  const _valid = assertGaugeValidity({ chainId, poolId, rewardTokenId, distributorId })
-
-  const curve = useStore.getState().curve
-  const pool = curve.getPool(_valid.poolId)
-  return pool.gauge.addReward(_valid.rewardTokenId, _valid.distributorId)
+export const mutateAddRewardToken = async ({ poolId, rewardTokenId, distributorId }: AddRewardQuery): Promise<string> => {
+  const { curve } = useStore.getState()
+  console.log('mutateAddRewardToken', poolId, rewardTokenId, distributorId)
```
Comment (ignromanov) [inline apps/main/src/entities/gauge/api/rewards-mutate.ts:17]: I don't think we need it

File: apps/main/src/entities/gauge/model/mutate-options.ts [hunk 4]
```diff
@@ -11,51 +11,51 @@
...
+} from './gauge-validation'
+import { gaugeKeys as keys } from './query-keys'
 
-export const getAddRewardTokenMutation = ({ chainId, poolId }: GaugeQueryParams) => ({
-  mutationFn: async ({ rewardTokenId, distributorId }: AddRewardParams) =>
-    api.mutateAddRewardToken(keys.addRewardToken({ chainId, poolId, rewardTokenId, distributorId })),
-  mutationKey: keys.addRewardToken({ chainId, poolId }),
+export const getAddRewardTokenMutation = (gauge: GaugeParams) => ({
+  mutationFn: async (params: AddRewardMutation) => {
+    console.log('AddRewardToken', params);
```
Comment (ignromanov) [inline apps/main/src/entities/gauge/model/mutate-options.ts:33]: This too

File: apps/main/src/entities/gauge/api/rewards-mutate.ts (hunk 3 above)
Comment (DanielSchiavini) [reply apps/main/src/entities/gauge/api/rewards-mutate.ts]: indeed, sorry for that

File: apps/main/src/entities/gauge/model/mutate-options.ts (hunk 4 above)
Comment (DanielSchiavini) [reply apps/main/src/entities/gauge/model/mutate-options.ts]: fixed!

**Review by ignromanov — APPROVED** [review]

**Review by Macket — APPROVED** [review]

**Pull request merged**